File: Corte 2/Taller 1/Algoritmo de Ordenamiento Go/Insercion.go
package main

import (
    "fmt"
    "math/rand"
    "time"
)

const N = 1000

func insertionSort(arr []int, comparisons *int, swaps *int) {
    for i := 1; i < N; i++ {
        key := arr[i]
        j := i - 1

        // Comparar e insertar
        for j >= 0 && arr[j] < key {
            (*comparisons)++
            arr[j+1] = arr[j]
            j--
            (*swaps)++
        }
        arr[j+1] = key
        (*swaps)++
    }
}

func main() {
    // Inicializar la semilla para números aleatorios
    rand.Seed(time.Now().UnixNano())

    // Generar números aleatorios entre 0 y 9
    arr := make([]int, N)
    for i := range arr {
        arr[i] = rand.Intn(10) // Números entre 0 y 9
    }

    // Mostrar números generados
    fmt.Println("Números generados:")
    for _, num := range arr {
        fmt.Printf("%d ", num)
    }
    fmt.Println()

    var comparisons int = 0
    var swaps int = 0

    // Ordenar usando el algoritmo de inserción
    insertionSort(arr, &comparisons, &swaps)

    // Mostrar números ordenados
    fmt.Println("Números ordenados:")
    for _, num := range arr {
        fmt.Printf("%d ", num)
    }
    fmt.Println()

   
    fmt.Printf("Comparaciones: %d\n", comparisons)
    fmt.Printf("Intercambios: %d\n", swaps)
    fmt.Println("Estable: Sí")
    fmt.Println("Inserción: Sí")
}

File: Corte 2/Taller 1/Algoritmo de Ordenamiento Python/insercion.py
import random
import time

# Función de ordenamiento por inserción
def insertion_sort(arr, comparisons, swaps):
    n = len(arr)
    for i in range(1, n):
        key = arr[i]
        j = i - 1

        # Comparar e insertar
        while j >= 0 and arr[j] > key:
            comparisons[0] += 1
            arr[j + 1] = arr[j]
            j -= 1
            swaps[0] += 1
        arr[j + 1] = key
        swaps[0] += 1

def main():
    N = 1000
    arr = [0] * N
    comparisons = [0]  # Usamos una lista para permitir la modificación dentro de la función
    swaps = [0]

    # Medir el tiempo de ejecución
    start_time = time.perf_counter_ns()

    # Generar números aleatorios de 0 a 9
    for i in range(N):
        arr[i] = random.randint(0, 9)  # Números entre 0 y 9

    # Mostrar números generados
    print("Números generados:")
    print(" ".join(map(str, arr)))

    # Ordenar usando el algoritmo de inserción
    insertion_sort(arr, comparisons, swaps)

    end_time = time.perf_counter_ns()

    # Calcular el tiempo transcurrido en nanosegundos
    time_taken = end_time - start_time

    # Mostrar números ordenados
    print("\nNúmeros ordenados:")
    print(" ".join(map(str, arr)))

    # Mostrar resultados
    print(f"\nComparaciones: {comparisons[0]}")
    print(f"Intercambios: {swaps[0]}")
    print("Estable: Sí")
    print("Inserción: Sí")

    # Mostrar el tiempo de ejecución en nanosegundos
    print(f"Tiempo de ejecución en nanosegundos: {time_taken}")

if __name__ == "__main__":
    main()
